### PR TITLE
add `internal-lints` feature to enable clippys internal lints (off by default)

### LINKED ISSUE #6306: Put internal lints behind a feature flag
We shouldn't ship internal lints to end users. It's just wasted bytes and late passes. There is also the possibility that an internal lint "escapes from the the lab" and causes crashes or bad suggestions. We can put the lints behind a feature flag and only enable it when doing the `dogfood` test. That would ensure that the lints continue to check Clippy code without making it into a release.

### PR DESCRIPTION
**Pull request by matthiaskrgr**
This PR moves the internal lint tests into a new subdirectory (I couldn't find a different way to compile-time-conditionally exclude them from compiletest) and only builds and tests internal lints if the `internal-lints` feature is enabled.

Fixes #6306

changelog: put internal lints behind a feature ("internal-lints")

### CODE REVIEW

**Comment by flip1995** [comment]
This will also require updating the `clippy_dev update_lints` file updates.

**Review by llogiq — CHANGES_REQUESTED** [review]
Just a small nit, and of course the clippy_dev update, but otherwise I like this change. Makes clippy smaller & faster to both build & run with no downside.

**Comment by flip1995** [comment]
This will also require some CI changes to enable internal lints in CI. I would like to review those (or do them myself, if you want) before this gets merged, but I also like this change generally speaking.

**Comment by matthiaskrgr** [comment]
Hm, the current roadblock is that `compiletest` does not seem to provide a way to skip tests based on crate features, but I need to skip some of the internal tests during default compilation...
https://rustc-dev-guide.rust-lang.org/tests/adding.html#other-header-commands

**Comment by matthiaskrgr** [comment]
Had the showerthought to put all the internal tests into a separate directory, that might make it work.

**Comment by matthiaskrgr** [comment]
I moved the internal tests into a new directory and `cargo test` and `cargo test --features internal-lints` passes now! :tada: 
`cargo dev update_lints` however tries to remove my `#[cfg(feature = "internal-lints")]` annotations again hmm.

**Comment by matthiaskrgr** [comment]
Ah this needs a rustup first... https://github.com/rust-lang/rust-clippy/pull/6329

File: .github/workflows/clippy_bors.yml
Comment (flip1995) [inline .github/workflows/clippy_bors.yml:139]: Couldn't we just directly build clippy with internal lints enabled?

File: clippy_lints/src/lib.rs
Comment (flip1995) [inline clippy_lints/src/lib.rs:1120]: It would be nice to have the internal lints all grouped in one block and put that block at the top of the register list.

File: tests/dogfood.rs
Comment (flip1995) [inline tests/dogfood.rs:30]: Since we enable `--all-features` for dogfood anyway, I don't think it makes a difference to also add `--features internal-lints`.

**Review by flip1995 — CHANGES_REQUESTED** [review]

**Comment by flip1995** [comment]
With this change, we can also remove those hacks:
https://github.com/rust-lang/rust-clippy/blob/ad4f82997a94cc91723daae14889f21428e65472/clippy_lints/src/utils/internal_lints.rs#L296-L298

Also we could move the `internal_lints` module out of utils and split it up into more modules, like the `methods` module. But this may be more involved and better be done in another PR.

File: tests/dogfood.rs
Comment (matthiaskrgr) [reply tests/dogfood.rs:30]: Ah right!

**Comment by matthiaskrgr** [comment]
(still need to take a closer look at `internal_lints.rs`)

File: clippy_lints/src/lib.rs
Comment (flip1995) [inline clippy_lints/src/lib.rs:957]: Please remove the empty lines.

File: tests/dogfood.rs
Comment (flip1995) [inline tests/dogfood.rs:21]: You can save the command in a variable and then reduce the if:

```rust
let mut command = Command::new(..)......;

if cfg!(feature = "internal-lints") {
    command.args(&["-D", "clippy::internal"]);
}

let output = command.output().unwrap();
```

Comment (matthiaskrgr) [reply tests/dogfood.rs:21]: Merged the changes into the last commit :+1:

File: clippy_lints/src/lib.rs
Comment (flip1995) [inline clippy_lints/src/lib.rs:1143]:
```suggestion
```

File: clippy_lints/src/lib.rs
Comment (flip1995) [inline clippy_lints/src/lib.rs:1157]:
```suggestion
```

File: clippy_lints/src/lib.rs
Comment (flip1995) [inline clippy_lints/src/lib.rs:1173]:
```suggestion
```

File: clippy_lints/src/lib.rs
Comment (flip1995) [inline clippy_lints/src/lib.rs:518]: Oh no, this'll get rid of our sole easteregg in Clippy :(

Comment (matthiaskrgr) [reply clippy_lints/src/lib.rs:518]: Mmh, is this bad?

File: clippy_dev/src/lib.rs
Comment (phansch) [inline clippy_dev/src/lib.rs:164]: nit: `extend` below should also work [with iterators](https://doc.rust-lang.org/std/iter/trait.Extend.html), so we can probably remove these `collect` calls?

Comment (matthiaskrgr) [reply clippy_dev/src/lib.rs:164]: neat! updated.

File: clippy_lints/src/lib.rs
Comment (flip1995) [reply clippy_lints/src/lib.rs:518]: No, just a bit sad 😄

**Review by flip1995 — APPROVED** [review]

**Comment by flip1995** [comment]
@bors r+

Thanks! 

(I dismissed @llogiq review, because it was about something already fixed)

**Pull request merged**
